### PR TITLE
First stage of RegUtil simplification

### PR DESCRIPTION
**Pull request by jbloom22**

### CODE REVIEW

**Comment by jbloom22** [comment]
@danking this is just the first step, but it provides a good interface to the function you need for PCrelate

File: python/hail/dataset.py [hunk 1]
```diff
@@ -2798,7 +2798,7 @@ def lmmreg(self, kinshipMatrix, y, covariates=[], global_root="global.lmmreg", v
         :param delta: Fixed delta value to use in the global model, overrides fitting delta.
         :type delta: float or None
 
-        :param float sparsity_threshold: AF threshold above which to use dense genotype vectors in rotation (advanced).
+        :param float sparsity_threshold: Genotype vector sparsity below which to use sparse genotype vector in rotation (advanced).
```
Comment (cseed) [inline python/hail/dataset.py]: at or below which

File: src/main/scala/is/hail/methods/LinearRegression.scala [hunk 2]
```diff
@@ -51,18 +51,22 @@ object LinearRegression {
...
-          RegressionUtils.toLinregHardCallStats(gs, yBc.value, sampleMaskBc.value, combinedMinAC)
-
-      val linregAnnot = optStats.map { case (x, xx, xy) =>
-
+      val (x: Vector[Double], isValid: Boolean) =
+        if (useDosages) {
+          val (x, mean) = RegressionUtils.dosageStats(gs, Some(sampleMaskBc.value), n)
+          (x, n * mean >= combinedMinAC)
+        }
+        else {
```
Comment (cseed) [inline src/main/scala/is/hail/methods/LinearRegression.scala]: formatting: move to previous line

File: src/main/scala/is/hail/methods/LinearRegressionMultiPheno.scala [hunk 3]
```diff
@@ -55,17 +55,21 @@ object LinearRegressionMultiPheno {
...
-          RegressionUtils.toLinregHardCallStats(gs, yDummyBc.value, sampleMaskBc.value, combinedMinAC)
-
-      val linregAnnot = optStats.map { case (x, xx, _) =>
-
+      val (x: Vector[Double], isValid: Boolean) =
+        if (useDosages) {
+          val (x, mean) = RegressionUtils.dosageStats(gs, Some(sampleMaskBc.value), n)
+          (x, n * mean >= combinedMinAC)
+        }
+        else {
```
Comment (cseed) [inline src/main/scala/is/hail/methods/LinearRegressionMultiPheno.scala]: Same.

File: src/main/scala/is/hail/stats/RegressionUtils.scala [hunk 4]
```diff
@@ -287,92 +287,71 @@ object RegressionUtils {
     }
   }
 
-  // constructs DenseVector of dosages (with missing values mean-imputed) in parallel with other statistics sufficient for linear regression
-  def toLinregDosageStats(gs: Iterable[Genotype], y: DenseVector[Double], mask: Array[Boolean], minAC: Int): Option[(DenseVector[Double], Double, Double)] = {
-    val nMaskedSamples = y.length
-    val valsX = Array.ofDim[Double](nMaskedSamples)
+  def dosageStats(gs: Iterable[Genotype], optMask: Option[Array[Boolean]], nKept: Int): (DenseVector[Double], Double) = {
+    val noMask = optMask.isEmpty
+    val mask = if (noMask) null else optMask.get
```
Comment (cseed) [inline src/main/scala/is/hail/stats/RegressionUtils.scala]: match opt `optMask`.  No using isEmpty/get.

Comment (jbloom22) [inline src/main/scala/is/hail/stats/RegressionUtils.scala]: I had this originally, is it what you have in mind?
```
    optMask match {
      case Some(mask) =>
        var i = 0
        while (gts.hasNext) {
          val gt = gts.next()
          if (mask(i)) {
            if (gt != -1) {
              valsX(row) = gt
              sumX += gt
            } else {
              nMissing += 1
              missingRowIndices += row
            }
            row += 1
          }
          i += 1
        }
      case None =>
        while (gts.hasNext) {
          val gt = gts.next()
          if (gt != -1) {
            valsX(row) = gt
            sumX += gt
          } else {
            nMissing += 1
            missingRowIndices += row
          }
          row += 1
        }
    }
```

**Comment by jbloom22** [comment]
Back to you

File: src/main/scala/is/hail/stats/RegressionUtils.scala [hunk 5]
```diff
@@ -287,92 +287,83 @@ object RegressionUtils {
...
-      }
-
-      val x = DenseVector(valsX)
-      val xx = sumXX + meanX * meanX * nMissing
-      val xy = sumXY + meanX * sumYMissing
+    val gts = gs.dosageIterator
 
-      Some(x, xx, xy)
+    optMask match {
+      case Some(mask) =>
```
Comment (cseed) [inline src/main/scala/is/hail/stats/RegressionUtils.scala]: I'd just make this one block and put the if inside the loop.

More meta comment, I'm slightly confused by these changes.  I thought you were going to let the user write an expression for the genotype-dependent covariate?

Comment (jbloom22) [inline src/main/scala/is/hail/stats/RegressionUtils.scala]: That's a related but separate set of changes. This arc is toward "de-optimizing" what was over-optimized to simplify the code base (remember we discussed the problem of quadratic growth in RegUtils?).  Plenty more code to delete.

**Comment by jbloom22** [comment]
Is `(optMask.isEmpty || optMask.get(i))` reasonable?

**Comment by cseed** [comment]
I'm basically happy with this or you can use null instead of Option, assuming, in light of our discussion, you still want it to go in.

**Comment by jbloom22** [comment]
I changed Option to null, let's merge as Dan will use hardCallStats right away

File: src/main/scala/is/hail/stats/RegressionUtils.scala [hunk 6]
```diff
@@ -287,92 +287,63 @@ object RegressionUtils {
...
-
+    var row = 0
     val gts = gs.dosageIterator
+
     var i = 0
-    var row = 0
     while (gts.hasNext) {
       val gt = gts.next()
-      if (mask(i)) {
+      if (mask == null || mask(i)) {
```
Comment (cseed) [inline src/main/scala/is/hail/stats/RegressionUtils.scala]: I don't see how mask can be null.

**Pull request merged**